src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java

```diff
@@ -159,6 +172,93 @@ public static EndTrackingResponse forciblyEndTracking(Request request) {
+            if (trackedJourney != null) {
+                if (rerouteTrip(tripData, false)) {
+                    return doUpdateTracking(request, tripData, false);
+                } else {
+                    return new TrackingResponse(
+                        TRIP_TRACKING_UPDATE_FREQUENCY_SECONDS,
+                        "No itinerary found!",
+                        trackedJourney.id,
+                        TripStatus.DEVIATED.name()
+                    );
+                }
+            } else {
+                logMessageAndHalt(request, HttpStatus.BAD_REQUEST_400, "Journey for provided trip id does not exist!");
+                return null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Attempt to reroute from the traveler's current location to the trip's original destination.
+     */
+    public static boolean rerouteTrip(TripTrackingData tripData, boolean isTest) {
+        IS_TEST = isTest;
+        if (tripData != null) {
+            var trackedJourney = tripData.journey;
+            return trackedJourney != null && updateTripWithNewItinerary(trackedJourney);
+        }
+        return false;
+    }
+
+    /**
+     * Record the location and time (as Strings to appease Mongo) a trip was rerouted and reset deviations.
+     */
+    private static void registerRerouting(TrackedJourney trackedJourney) {
+        trackedJourney.longestConsecutiveDeviatedPoints = -1;
+        trackedJourney.reroutings.put(
+            new Coordinates(trackedJourney.lastLocation()).getCoordinates(),
+            LocalDateTime.now().toString()
+        );
+        Persistence.trackedJourneys.replace(trackedJourney.id, trackedJourney);
+    }
+
+    /**
+     * Retrieve a new itinerary from OTP and update the associated trip's matching itinerary.
+     */
+    private static boolean updateTripWithNewItinerary(TrackedJourney trackedJourney) {
+        Itinerary itinerary = getItineraryFromOtpResponse(trackedJourney);
+        if (itinerary != null) {
+            trackedJourney.trip.journeyState.matchingItinerary = itinerary;
+            Persistence.monitoredTrips.replace(trackedJourney.trip.id, trackedJourney.trip);
+            registerRerouting(trackedJourney);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get the itinerary with the shortest duration returned from OTP using the new start location and current time.
+     */
+    private static Itinerary getItineraryFromOtpResponse(TrackedJourney trackedJourney) {
+        if (IS_TEST) {
+            // return the original trip itinerary which for testing purposes will differ from the matching itinerary.
+            return trackedJourney.trip.itinerary;
+        }
+        try {
+            OtpGraphQLVariables query = trackedJourney.trip.otp2QueryParams;
+            query.fromPlace = new Coordinates(trackedJourney.lastLocation()).getCoordinates();
+            query.time = getTimeNowAsString();
```

Comment:
Extract a method and write unit tests for it.

Reply:
Addressed in: https://github.com/ibi-group/otp-middleware/pull/286/commits/23d04728e3e4d229384d0e37b6ca821b9f416208